builtin/providers/vsphere/resource_vsphere_folder.go

```diff
@@ -0,0 +1,229 @@
+	path              string
+}
+
+func resourceVSphereFolder() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceVSphereFolderCreate,
+		Read:   resourceVSphereFolderRead,
+		Delete: resourceVSphereFolderDelete,
+
+		Schema: map[string]*schema.Schema{
+			"datacenter": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"path": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"existing_path": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceVSphereFolderCreate(d *schema.ResourceData, meta interface{}) error {
+	
+	client := meta.(*govmomi.Client)
+
+	f := folder{
+		path: strings.TrimRight(d.Get("path").(string), "/"),
+	}
+
+	if v, ok := d.GetOk("datacenter"); ok {
+		f.datacenter = v.(string)
+	}
+
+	createFolder(client, &f)
+
+	d.Set("existing_path", f.existingPath)
+	d.SetId(fmt.Sprintf("%v/%v", f.datacenter, f.path))
+	log.Printf("[INFO] Created folder: %s", f.path)
+
+	return resourceVSphereFolderRead(d, meta)
+}
+
+
+func createFolder(client *govmomi.Client, f *folder) error {
+
+	finder := find.NewFinder(client.Client, true)
+
+	dc, err := finder.Datacenter(context.TODO(), f.datacenter)
+	if err != nil {
+		return fmt.Errorf("error %s", err)
+	}
+	finder = finder.SetDatacenter(dc)
+	si := object.NewSearchIndex(client.Client)
+
+	dcFolders, err := dc.Folders(context.TODO())
+	if err != nil {
+		return fmt.Errorf("error %s", err)
+	}
+
+	folder := dcFolders.VmFolder
+	var workingPath string
+
+	pathParts := strings.Split(f.path, "/")
+	for _, pathPart := range pathParts {
+		if len(workingPath) > 0 {
+			workingPath += "/"
+		}
+		workingPath += pathPart
+		subfolder, err := si.FindByInventoryPath(
+			context.TODO(), fmt.Sprintf("%v/vm/%v", f.datacenter, workingPath))
+
+ 		if err != nil {
+ 			return fmt.Errorf("error %s", err)
+ 		} else if subfolder == nil {
+ 			log.Printf("[DEBUG] folder not found; creating: %s", workingPath)
+ 			folder, err = folder.CreateFolder(context.TODO(), pathPart)
+			if err != nil {
+				return fmt.Errorf("Failed to create folder at %s; %s", workingPath, err)
+			}
+ 		} else {
+ 			log.Printf("[DEBUG] folder already exists: %s", workingPath)
+ 			f.existingPath = workingPath
+ 			folder = subfolder.(*object.Folder)
+ 		}
+	}
+	return nil
+}
+
+
+func resourceVSphereFolderRead(d *schema.ResourceData, meta interface{}) error {
+	
+	log.Printf("[DEBUG] reading folder: %#v", d)
+	client := meta.(*govmomi.Client)
+	
+	dc, err := getDatacenter(client, d.Get("datacenter").(string))
+	if err != nil {
+		return err
+	}
+	
+	finder := find.NewFinder(client.Client, true)
+	finder = finder.SetDatacenter(dc)
+
+	folder, err := object.NewSearchIndex(client.Client).FindByInventoryPath(
+			context.TODO(), fmt.Sprintf("%v/vm/%v", d.Get("datacenter").(string), 
+				d.Get("path").(string)))
+
+	if err != nil || folder == nil {
```

Comment:
Is there any way to scope this error check down to a "NotFound" error and return the error in other cases?

If we treat any error here as though the resource was deleted, running a `terraform plan` with e.g. internet disconnected could result in state losing track of all its folders instead of the error message you'd expect in that case.